### PR TITLE
feat: NetworkObjectSpawner for dynamically spawning in-scene placed NetworkObjects [MTT-1558]

### PR DESCRIPTION
**Pull request by fernando-cortez**
### Description
This PR resolves the issue encountered when joining a game session already underway, where the late-joiner would see ghost imps. In-scene placed NetworkObjects were destroyed & despawned on the server, and so the joining client would always load new NetworkObjects which were not synchronized with the server.

Here, `NetworkObjectSpawner` component can be added to a GameObject inside a scene/prefab to spawn a NetworkObject when NetworkSceneManager's OnLoadEventCompleted is invoked. The wrapper is destroyed after the NetworkObject is spawned inside the server, and otherwise on Awake() on clients, to reduce memory buildup.

Gizmos have been added as a way to at-a-glance spot NetworkObjects in a scene/prefab that will be spawned. I would love to add meshes as a preview, but in my opinion that's more of a nice-to-have, cosmetic, editor tool.

An extra: VandalImpGraphics had an unnecessary NetworkObject added to it which prevented spawns on the entrance area. This has been removed.

### Issue Number(s)
[MTT-1558](https://jira.unity3d.com/browse/MTT-1558)

### Contribution checklist
 - [ ] Tests have been added for boss room and/or utilities pack
 - [ ] Release notes have been added to the [project changelog](../CHANGELOG.md) file and/or [package changelog](../Packages/com.unity.multiplayer.samples.coop/CHANGELOG.md) file
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] JIRA ticket ID is in the PR title or at least one commit message
 - [ ] Include the ticket ID number within the body message of the PR to create a hyperlink

### CODE REVIEW

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+using UnityEngine;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    /// <summary>
+    /// Custom spawning component to be added to a scene GameObject. This component collects NetworkObjects in a scene
+    /// marked by a special tag, collects their Transform data, destroys their prefab instance, and performs the dynamic
+    /// spawning of said objects during Netcode for GameObject's (Netcode) OnNetworkSpawn() callback.
+    /// </summary>
+    [RequireComponent(typeof(NetcodeHooks))]
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:17]: Are there chances this object will get its client or server version stripped?
Why the Netcode hooks here?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                spawnedGameObject.transform.localScale = m_SpawnObjectData[i].transform.lossyScale;
+                var spawnedNetworkObject = spawnedGameObject.GetComponent<NetworkObject>();
+                spawnedNetworkObject.Spawn();
+
+                Destroy(m_SpawnObjectData[i].gameObject);
+            }
+        }
+
+#if UNITY_EDITOR
+        public void CollectTaggedPrefabInstances()
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:80]: I'm confused, didn't we say we'd have one spawner instance per static instance? And so we'd have 100 spawners in scene, for each placed imp? And this way, each spawner could reference the imp prefab to spawn and no need for all this logic to maintain?

Comment (fernando-cortez) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:80]: This is the spawner object you'd ideally drop into your scene that already contained NetworkObjects, and there'd be only one of these ideally. We don't have spawner instances, but just spawner data GameObjects, such that we're still able to move/rotate transforms after converting them into data GameObjects.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs (hunk 1 above)
Comment (fernando-cortez) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:17]: The spawning logic should only operate on the server, and we need NetcodeHooks to know when this in-scene placed NetworkObject is spawned by the network.

Would renaming `ServerNetworkObjectSpawner` help here?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs (hunk 2 above)
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:80]: we'd still have all this extra logic to maintain no? why the complexity?

Comment (fernando-cortez) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:80]: I kept this around initially to help myself convert the prefab instances into the data gameobjects, but if we don't wish to keep this around, we can enforce dropping the data NetworkObject prefab into scenes as the only flow for spawning in-scene placed networkobjects.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs (hunk 1 above)
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:17]: what I meant is why not have this be a NetworkBehaviour directly?

Comment (fernando-cortez) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:17]: That is a good question. There shouldn't be any other MonoBehaviours to listen for callbacks. Will refactor.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs (hunk 2 above)
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:80]: the less code we need to maintain and debug, the better.
the end goal is to have something that's easy to use by a game designer. the simpler the better!

**Review by SamuelBellomo — APPROVED** [review]

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/SpawnObjectData.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+using UnityEngine;
+
+namespace Unity.Multiplayer.Samples.Utilities
+{
+    public class SpawnObjectData : MonoBehaviour
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/Net/SpawnObjectData.cs:5]: this could still be merged right? or am I reviewing too early? :P

Comment (fernando-cortez) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/SpawnObjectData.cs:5]: I've got no more modifications I wish to make, so yeah 😄

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/Net/SpawnObjectData.cs:5]: So in this case, yeah, that data could be merged with the spawner (and no need for list) --> it'd make this even simpler

**Review by LPLafontaineB — COMMENTED** [review]
Really elegant solution! Something that might be an issue, though, is that now the dynamically-spawned imps are no longer linked to their respective additive scene. So if we move out of it and it gets unloaded, the imps still stay there, but if we come back to it and it gets loaded again, the imps get re-spawned. This means we can have multiple bosses or vandal imps at once

File: Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs [hunk 4]
```diff
@@ -3,52 +3,47 @@
...
-                var spawnedGameObject = Instantiate(m_SpawnObjectData[i].prefabReference,
-                    m_SpawnObjectData[i].transform.position,
-                    m_SpawnObjectData[i].transform.rotation,
-                    null);
-
-                spawnedGameObject.transform.localScale = m_SpawnObjectData[i].transform.lossyScale;
-                var spawnedNetworkObject = spawnedGameObject.GetComponent<NetworkObject>();
-                spawnedNetworkObject.Spawn();
+            SpawnNetworkObject();
+            Destroy(gameObject);
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/Net/NetworkObjectSpawner.cs:39]: nice, I like that cleanup

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,7 +11,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Added
 * Added handling the OnTransportFailure callback (#707). This callback is invoked when a failure happens on the transport's side, for example if the host loses connection to the Relay service. This won't get called when the host is just listening with direct IP, this would need to be handled differently (by pinging an external service like google to test for internet connectivity for example). Boss Room now handles that callback by returning to the Offline state.
 * Pickup and Drop action added to the Action system. Actionable once targeting a "Heavy"-tagged NetworkObject. (#372) - This shows NetworkObject parenting with a pattern to follow animation bones (the hands when picking up)
-*
+* NetworkObjectSpawner handles dynamically spawning in-scene placed NetworkObjects (#717) - This showcases proper handling of NetworkObjects that you'd wish to place inside of scenes, but would still want to destroy at game-time. Examples of these are: Imps, VandalImps, ImpBoss.
```
Comment (SamuelBellomo) [inline CHANGELOG.md:14]:
```suggestion
* NetworkObjectSpawner handles dynamically spawning in-scene placed NetworkObjects (#717) - You can't place a NetworkObject in scene directly and destroy it at runtime. This PR showcases proper handling of NetworkObjects that you'd wish to place inside of scenes, but would still want to destroy at game-time. Examples of these are: Imps, VandalImps, ImpBoss. Doors, buttons, cristals, etc stay the same, statically placed in scene.
```

**Review by SamuelBellomo — DISMISSED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by LPLafontaineB — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by LPLafontaineB — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Pull request merged**